### ISSUE TITLE
Improve the error message for passing `(any P)?` to `(some P)?`

### ISSUE THREAD

**Issue by hborla** [apple/swift#61733]
If you try to pass a value of type `(any P)?` to an argument of type `(some P)?`, the compiler tells you that `any P` cannot conform to `P`, which is confusing and not actionable. With [SE-0375](https://github.com/apple/swift-evolution/blob/main/proposals/0375-opening-existential-optional.md), existential opening/unboxing is supported with optionals, but only if the argument is not an optional because it's possible that the `(any P)?` is `nil` (and thus does not have a dynamic type to bind to `some P`). With SE-0375, the fix is to unwrap the optional or provide a default value. So, the error message should tell you that!

```swift
protocol P {}

struct S: P {}

func takesOptionalP(_: (some P)?) {}

func passOptional(value: (any P)?) {
  takesOptionalP(value) // error: type 'any P' cannot conform to 'P'

  takesOptionalP(value ?? S()) // okay under SE-0375
  takesOptionalP(value!) // okay under SE-0375
}
```

The constraint system currently applies the `MissingConformance` fix when solving for the `takesOptionalP(value)` expression. Instead, it should identify that the wrapped type of the argument matches the parameter type, and apply the `ForceOptional` fix.

**Comment by hborla** [comment]
@dfperry5 this is very similar to https://github.com/apple/swift/issues/60730 in case you're interested in taking a look at this one too 🙂

**Comment by IceCurrent** [comment]
@hborla, I would love to work on it. But I'm a completely new to this repo (and open source contribution in general). It would be very kind of you tell me how do I go about it

**Comment by hborla** [comment]
@IceCurrent welcome to the Swift project! I'm happy you're interested in working on this issue 🙂

Are you set up to build the Swift compiler yet? There are a few getting started guides:
* [How to Set Up an Edit-Build-Test-Debug Loop](https://github.com/apple/swift/blob/main/docs/HowToGuides/GettingStarted.md)
* [How to Submit Your First Pull Request](https://github.com/apple/swift/blob/main/docs/HowToGuides/FirstPullRequest.md)

The first one contains instructions for checkout and build the Swift compiler, and the second has an FAQ for new contributors. Let me know once you've gotten set up and I can provide some guidance on investigating this issue!

**Comment by enathang** [comment]
@IceCurrent Hey! Are you still working on this? 

I can take this fix up if you're no longer interested.

**Comment by IceCurrent** [comment]
@hborla I'm really sorry I could not work on it right now. There has been some family emergency, which I'm busy dealing with.     @enathang It would be great if you could take up this issue. 
Sorry once again!

**Comment by hborla** [comment]
No worries at all! @enathang please let me know if you have questions or need guidance!

**Comment by enathang** [comment]
Thanks @hborla! Currently waiting on a new external hard drive because my computer ran out of space trying to build the project. In the mean time, I looked over the documentation (specifically [the overview of constraint system](https://www.swift.org/blog/new-diagnostic-arch-overview/)) and [this](https://github.com/apple/swift/pull/61191/files) commit. If you have any additional info on the constraint system feel free to pass along, it seems cool! 

Will update once I have the project built and start poking through files.

**Comment by enathang** [comment]
@hborla Is there any info on reducing the size requirement of Swift/xcode/swift toolchain? I'm wondering if I'd like to only work on a part of the toolchain (ie type constraint) I could swap out other parts of the toolchain with compiled versions to reduce space. 

I have a 2017 128GB macbook air. I've stripped out every non-swift related file on my computer to free space and I'm still running into disk space issues. Your thoughts would be much appreciated!

**Comment by enathang** [comment]
Managed to get everything working through a flash drive. 

I've done a bit of testing to determine the behavior of the compiler currently:
Passing `any P` to `some P` -> No error
Passing `any P` to `(some P)?` -> No error

Passing `(any P)?` to `(some P)?` -> NonConformance 
Passing `(any P)?` to `some P` -> Both NonConformance and Unwrap literal

I'll keep looking into this.

**Comment by NuriAmari** [comment]
@enathang Are you still working on this? If not, I'd like to give it a try.

**Comment by enathang** [comment]
Go for it! I worked on it for a while but never quite figured it out.

**Comment by NuriAmari** [comment]
Hi @LucianoPAlmeida @hborla @AnthonyLatsis, I've given this a try but could use some pointers if someone has the time.

The crux of the issue I'm running into is convincing the type checker to choose the `ForceOptional` solution over the `MissingConformance` solution. The change introduced in https://github.com/apple/swift/pull/61191 already makes it so that the `ForceOptional` solution is considered, but the solver bails before fully considering the solution because it is strictly worse than the chosen `MissingConformance` solution, due to the `value-to-optional` promotion penalty.

We can see this in the constraint debugging output:

```
---Attempting to salvage and emit diagnostics---
  (considering -> $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];
    (simplification result:
    )
    (outcome: unsolved)
  )
  (attempting disjunction choice (P)? bind ($T1)? [deep equality] [[locator@0x11d9ac9f0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0]]];
    (considering -> $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];
      (simplification result:
    (attempting fix [fix: add missing protocol conformance] @ locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)])
    (increasing 'applied fix' score by 1)
        (removed constraint: $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];)
      )
      (outcome: simplified)
    )
    (Changes:
      (Newly Bound:
        > $T1 := (P)
      )
      (Removed Constraint:
        > $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];
      )
    )
    (found solution: [component: applied fix(s), value: 1])
  )
  (attempting disjunction choice (P)? arg conv ($T1)? [optional-to-optional] [[locator@0x11d9ac9f0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0]]];
        (added constraint: (P) arg conv $T1 [[locator@0x11d9acc40 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> generic argument #0]]];)
    (Changes:
      (Added Constraint:
        > (P) arg conv $T1 [[locator@0x11d9acc40 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> generic argument #0]]];
      )
    )
    (Potential Binding(s):
      ($T1 [attributes: potentially_incomplete] [with possible bindings: (supertypes of) P])
    )
    (attempting type variable $T1 := P
      (considering -> $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];
        (simplification result:
      (attempting fix [fix: add missing protocol conformance] @ locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)])
      (increasing 'applied fix' score by 1)
          (removed constraint: $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];)
        )
        (outcome: simplified)
      )
      (considering -> (P) arg conv $T1 [[locator@0x11d9acc40 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> generic argument #0]]];
        (simplification result:
          (removed constraint: (P) arg conv $T1 [[locator@0x11d9acc40 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> generic argument #0]]];)
        )
        (outcome: simplified)
      )
      (Changes:
        (Newly Bound:
          > $T1 := P
        )
        (Removed Constraints:
          > $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];
          > (P) arg conv $T1 [[locator@0x11d9acc40 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> generic argument #0]]];
        )
      )
      (found solution: [component: applied fix(s), value: 1])
    )
  )
  (attempting disjunction choice (P)? arg conv ($T1)? [value-to-optional] [[locator@0x11d9ac9f0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0]]];
    (increasing 'value to optional promotion' score by 1)
        (added constraint: (P)? arg conv $T1 [[locator@0x11d9accf0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> optional payload]]];)
    (Changes:
      (Added Constraint:
        > (P)? arg conv $T1 [[locator@0x11d9accf0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> optional payload]]];
      )
    )
    (Potential Binding(s):
      ($T1 [attributes: potentially_incomplete] [with possible bindings: (supertypes of) (P)?])
    )
    (attempting type variable $T1 := (P)?
      (considering -> $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];
        (simplification result:
      (attempting fix [fix: add missing protocol conformance] @ locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)])
      (increasing 'applied fix' score by 1)
      (solution is worse than the best solution)
      (attempting fix [fix: add missing protocol conformance] @ locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)])
      (increasing 'applied fix' score by 3)
      (solution is worse than the best solution)
          (removed constraint: $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];)
          (failed constraint $T1 conforms to P [[locator@0x11d9ac6d8 [DeclRef@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> opened generic -> type parameter requirement #0 (conformance)]]];)
        )
        (outcome: error)
      )
    )
    (attempting type variable $T1 := (P)
      (considering -> (P)? arg conv $T1 [[locator@0x11d9accf0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> optional payload]]];
        (simplification result:
      (attempting fix [fix: force optional] @ locator@0x11d9accf0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> optional payload])
      (increasing 'applied fix' score by 1)
      (solution is worse than the best solution)                                                                      !!!!!!!!!!             :(          !!!!!!!!!!!!!!!!!!!
          (removed constraint: (P)? arg conv $T1 [[locator@0x11d9accf0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> optional payload]]];)
          (failed constraint (P)? arg conv $T1 [[locator@0x11d9accf0 [Call@/Users/nuriamari/git/swift-project/swift/test/Constraints/scratch.swift:10:3 -> apply argument -> comparing call argument #0 to parameter #0 -> optional payload]]];)
        )
        (outcome: error)
      )
    )
  )
```

Strategies I've tried so far:

- Reduce the impact of the `ForceOptional` fix from 1 to 0
    - At least convinced the solver not to bail early, but it still required other fixes and seems like a hack with unwanted side effects
- Preventing the consideration of `MissingConformance` in the first place.
    - I found that in the locations where binding type variable T1 to concrete type P was being considered, I didn't have access to enough context to know that an `(any P)?` -> `(some P)?` conversion was being attempted.
    
Some general questions I have, again if there is time:

- In the debug output, I see "active constraints" and "inactive" constraints
    - What makes a constraint active vs inactive?
    - Perhaps related, how does the typechecker pick a starting constraint to begin simplifying. From the documentation I understand it computes connected components in the constraint graph, and independently solves each component, but its not clear to me how it picks which constraint within a component to start with. 
- How closely does the typechecker resemble Hindley-Milner? Would it be worth it / recommended to read some of the literature around it.

**Comment by LucianoPAlmeida** [comment]
> Preventing the consideration of MissingConformance in the first place.
I found that in the locations where binding type variable T1 to concrete type P was being considered, I didn't have access to enough context to know that an (any P)? -> (some P)? conversion was being attempted.

@hborla or @xedin may be the best advice here, but from what I could get from looking at it is that we are recording a missing conformance from an existential `any P` and protocol `P` (with an additional optional-to-optional or deep-equality restriction) which doesn't look correct. 
So maybe we should detect that existential case in `fixRequirementFailure` taking into account the restrictions as well and produce different fix. 

Also @hborla, it looks like the second case still fails https://godbolt.org/z/bzeP1EKM5. Is that a bug?

**Comment by Rajveer100** [comment]
This issue seems really interesting to grind! Let me try and find a solution to it!

**Comment by LucianoPAlmeida** [comment]
> This issue seems really interesting to grind! Let me try and find a solution to it!

You definitely can, but I think this issue has an assignee already. So we always recommend to ask the current assignee (or person who commented that is working on it) first. See [How do I pick something to work on](https://github.com/apple/swift/blob/main/docs/HowToGuides/FirstPullRequest.md#how-do-i-pick-something-to-work-on) for more details.

**Comment by NuriAmari** [comment]
> This issue seems really interesting to grind! Let me try and find a solution to it!

I'll still be working on it, but I have other things on my plate as well, so if you find a solution first, that's fine too 🙂

**Comment by Rajveer100** [comment]
Cool!

**Comment by Rajveer100** [comment]
Well, I am grinding as of now!

**Comment by Rajveer100** [comment]
So the fixit should be something like this right?

```
NOTE(did_you_mean_generic_param_as_conformance,none,
     "did you mean to declare %0 as a protocol conformance for %1?", (DeclName, Type))
```

**Comment by Rajveer100** [comment]
I think changes in these files could affect the error right?

```
lib/Sema/CSDiagnostics.cpp:  emitDiagnostic(diag::type_cannot_conform,
lib/Sema/CSDiagnostics.cpp:        emitDiagnostic(diag::type_cannot_conform, fromType, toType);
lib/Sema/TypeCheckProtocol.cpp:      diags.diagnose(ComplainLoc, diag::type_cannot_conform,
lib/Sema/TypeCheckProtocol.cpp:          diags.diagnose(ComplainLoc, diag::type_cannot_conform_to_nsobject,
```

**Comment by Rajveer100** [comment]
Can anyone briefly explain `Existential Types`, although I have looked up a little, need little more explanation, as there seems to be lot of checking for this in the above function I mentioned.

**Comment by NuriAmari** [comment]
> I think changes in these files could affect the error right?

In my opinion, you should be looking farther from the the diagnostic emission call sites, around the `CSSimpify.cpp` and `CSSolver.cpp` area off the top of my head. We need to modify which fixes the constraint system applies, not how it emits diagnostics for them after having decided which to apply. More concretely, we need to stop applying a conformance failure fix, not modify how a conformance failure is diagnosed 🙂 

Random collection of resources I used to get some understanding of existential types:

https://developer.apple.com/videos/play/wwdc2016/416/
https://developer.apple.com/videos/play/wwdc2022/110352
https://github.com/apple/swift-evolution/blob/main/proposals/0352-implicit-open-existentials.md#when-can-we-open-an-existential
https://github.com/apple/swift-evolution/blob/main/proposals/0375-opening-existential-optional.md
https://forums.swift.org/t/very-confused-about-some-compiler-warnings-related-to-protocols-bugs-or-intended/39754
https://forums.swift.org/t/allowing-self-conformance-for-protocols/39841

**Comment by Rajveer100** [comment]
Thanks! @NuriAmari

**Comment by aeu** [comment]
Pull request [86370](https://github.com/swiftlang/swift/pull/86370) created which fixes this.

**Comment by aeu** [comment]
Reviewer feedback incorporated, new pr [86773](https://github.com/swiftlang/swift/pull/86773) fixes this.

**Comment by al45tair** [comment]
Given
```
protocol P {}

struct S: P {}

func takesOptionalP(_: (some P)?) {}

func passOptional(value: (any P)?) {
  takesOptionalP(value)
}
```
it does seem, at first glance, odd from a language _user_'s perspective that `passOptional(value: nil)` wouldn't just result in `takesOptionalP(nil)`, though I appreciate that the problem is that we don't know what type `some P` is in that case (and we need that information in order to call `takesOptionalP` — it's just that that fact isn't immediately obvious from the language syntax).

I wonder if it might not be better to have the compiler generate a `no P` type that notionally conforms to `P` but that can't actually be used except as part of an `Optional` that is `nil`, so that

```
passOptional(value: nil)
```

results in a call to

```
takesOptionalP(nil as (no P)?)
```

It would be a compiler error to attempt to make use of any protocol method or property on a `no P` type, or to attempt to have a value of `(no P)?` that is non-`nil`.

(I filed this idea as rdar://171800406.)

**Comment by xedin** [comment]
I'm reverting the change here - https://github.com/swiftlang/swift/pull/88071
